### PR TITLE
Use cargo-check-external-types to control type leakage in public API

### PR DESCRIPTION
**Pull request by ctz**
I learned about this today, so thought I'd give it a try. It pointed out one `rustls-webpki` and two `ring` leakages that this PR fixes.

See https://github.com/rustls/rustls/actions/runs/6511651812/job/17687753330 for the example failing output.

### CODE REVIEW

**Review by cpu — APPROVED** [review]
Nice! This is super handy :rocket:

**Comment by ctz** [comment]
Working with upstream to get this going with a more recent nightly.

**Review by cpu — COMMENTED** [review]
Looks like this is working nicely with up-to-date nightly thanks to your upstream contrib. Cool!

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -312,3 +312,20 @@ jobs:
       - run: cargo clippy --locked --manifest-path=connect-tests/Cargo.toml --all-features --all-targets
       - run: cargo clippy --locked --manifest-path=fuzz/Cargo.toml --all-features --all-targets
       - run: cargo clippy --locked --manifest-path=provider-example/Cargo.toml --all-features --all-targets
+
+  check-external-types:
+    name: Validate external types apprearing in public API
```
[the remark this answers is not in the document]
Comment (cpu) [reply .github/workflows/build.yml:317]: Unfortunately looks like there's [a separate issue](https://github.com/awslabs/cargo-check-external-types/issues/113) that will prevent using this for rustls-webpki at the moment. I'll see about fixing that once I have some sense of what the best fix is.

[the remark this answers is not in the document]
Comment (ctz) [reply .github/workflows/build.yml]: Ah, good spot!

**Pull request merged**